stablehlo/transforms/StablehloAggressiveSimplificationPatterns.td

```diff
@@ -0,0 +1,281 @@
+// Copyright 2020 The IREE Authors
+//
+// Licensed under the Apache License, Version 2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// This is the legalization pattern definition file for CHLO to StableHLO.
+// These are included in the populateDecompositionPatterns factory
+// and should only include canonical expansions which are not actually
+// ambiguous/different for various backends. Avoid patterns that are actually
+// lowering to non-canonical forms.
+
+include "mlir/IR/OpBase.td"
+include "stablehlo/dialect/StablehloOps.td"
+
+//// Utilities
+def NotConstantOp : Constraint<
+    CPred<"llvm::isa<BlockArgument>($0) || !llvm::isa<stablehlo::ConstantOp>($0.getDefiningOp())">,
+    "is not a constant.">;
+
+def OperandsEqual : Constraint<CPred<"$0 == $1">, "operands are equal">;
+
+def TypesEqual : Constraint<CPred<"$0.getType() == $1.getType()">, "operands are equal">;
+
+def NumberOfElementsEqual : Constraint<
```

Comment:
Should we also consider the presence of dynamic dimensions here? Or maybe this simplification is applied after making sure of that dimensions are static.

Reply:
Simplification is not run after dims are static, but for patterns where this matters, we should be already using `AnyStaticShapeTensor` such as:

```
// Pattern: subtract(X, X) -> 0
// Must be static shape, otherwise would require broadcasting via CHLO_ConstantLike
def : Pat<(StableHLO_SubtractOp AnyStaticShapeTensor:$operand, $operand),
          (StableHLO_ConstantLike<"0"> $operand)>;
```